streaming/base/dataset.py

```diff
@@ -205,37 +206,99 @@ def _load_shard(self, shard: int, partition: Partition) -> None:
         """
         self._load_shards([shard], partition)
 
-    def _preload_shard(self, shard: int) -> bool:
-        """Decompress and validate a single shard, returning whether present.
+    def _decompress_shard_part(self, zip_info: FileInfo, zip_filename: str, raw_filename: str,
+                               compression: Optional[str], wait: bool):
+        """Validate and decompress shard data.
 
         Args:
-            shard (int): Which shard.
+            zip_info (FileInfo): Compressed file info.
+            zip_filename (str): Compressed filename.
+            raw_filename (str): Decompressed filename.
+            compression (str, optional): Compression algorithm.
+            wait (bool): Whether to wait for another worker to do the work.
+        """
+        # Doer or waiter?
+        if not wait:
+            # If doer: load compressed, validate, decompress, save decompressed, maybe remove
+            # compressed to save space.
+            data = open(zip_filename, 'rb').read()
+            if self.hash:
+                assert get_hash(self.hash, data) == zip_info.hashes[self.hash]
+            data = decompress(compression, data)  # pyright: ignore
+            tmp_filename = raw_filename + '.tmp'
```

Comment:
Wondering why to write first in `*.tmp` file and then rename it to a writer extension such as `mds` ? Why not writing directly to `*.mds` file?

Reply:
because we are doing a file existence check to determine when it is okay to proceed in the "waiter" process, the file must be entirely written as soon as it exists, so we write it to a temp file on the same filesystem then rename in place to the final name

Reply:
in a future PR, we should check file size as well as existence (but hash validation is too slow in that loop), as well as validating file size everywhere hashes are being compared, but let's treat that in its own PR

Reply:
Sounds Good. Make Sense.